controller/source/logging/logger.hpp

```diff
@@ -98,16 +99,77 @@ class logger {
               << (new_val.empty() ? "" : new_val) << std::endl;
   }
 
+  /*
+   * Logs the encoded query
+   */
+  void log_encoded_query(const query& query_args, const default_policy& def_policy, 
+                         const bool& valid, const std::string& new_val = {}) 
+  {  
+    // lock the mutex corresponding to the key
+    std::lock_guard<std::mutex> lock(m_keys_to_mutexes[query_args.key()]);
+
+    // Open or retrieve the file the file
+    auto log_file = get_or_open_log_stream(query_args.key());
 
-private:
-  logger() {
-    if (!std::filesystem::exists(logs_dir)) {
-      std::filesystem::create_directory(logs_dir);
+    // Encode the timestamp as a fixed-width integer type
+    const int64_t timestamp = std::chrono::system_clock::now().time_since_epoch().count();
+    // Encode the user key as a string
+    const std::string& user_key = query_args.user_key().value_or(def_policy.user_key());
+    // Encode the operation type (3bits) and the operation result (1 bit) as a single byte
+    const uint8_t operation = static_cast<uint8_t>((convert_operation_to_enum(query_args.cmd()) & 0x07U) << 1U);
+    const uint8_t valid_bit = (valid ? 0x01U : 0x00U);
+    const uint8_t operation_result = operation | valid_bit;
+    // Calculate the total size of the entry
+    // We need the size of the data + 3 delimiters + a new line char
+    size_t total_size = sizeof(timestamp) + user_key.length() + 
+                                sizeof(operation_result) + new_val.length() + 
+                                3 * sizeof(log_delimiter) + 1;
+
+    // Allocate an array buffer to hold the encoded entry
+    std::vector<char> buffer(total_size);
+    size_t offset = 0;
+
+    // Check if buffer is null
+    if (buffer.data() == nullptr) {
+      return;
     }
```

Comment:
Just curious - in what situations this can happen? 

Is the intention here to check if the allocation of a vector with the given size succeeded or not? If so, I think we get a bad_alloc error directly in the constructor.

Reply:
This was added to suppress a compiler warning for potential null-pointer dereference.
Enhanced with the `[[unlikely]]` hint for the compiler.